sql/query/expr.go

```diff
@@ -0,0 +1,660 @@
+		case stringOp:
+			v = *left.StringVal != *right.StringVal
+		}
+
+	case parser.In, parser.NotIn, parser.Like, parser.NotLike:
+		return null, fmt.Errorf("TODO(pmattis): unsupported comparison operator: %s", op)
+	}
+
+	return boolToDatum(v), nil
+}
+
+// Prepare the arguments for a binary operation. The returned arguments will
+// have the same type.
+func prepareBinaryArgs(typ opType, left, right sqlwire.Datum) (opType, sqlwire.Datum, sqlwire.Datum, error) {
+	var err error
+
+	switch typ {
+	case intOp, uintOp:
+		if left.UintVal != nil || right.UintVal != nil {
+			left, err = left.ToUint()
+			if err != nil {
+				return uintOp, null, null, err
+			}
+			right, err = right.ToUint()
+			if err != nil {
+				return uintOp, null, null, err
+			}
+			return uintOp, left, right, nil
+		}
+		left, err = left.ToInt()
+		if err != nil {
+			return intOp, null, null, err
+		}
+		right, err = right.ToInt()
+		if err != nil {
+			return intOp, null, null, err
+		}
+		return intOp, left, right, nil
+
+	case floatOp:
+		if (left.UintVal != nil && (right.IntVal != nil || right.UintVal != nil)) ||
+			(right.UintVal != nil && (left.IntVal != nil || left.UintVal != nil)) {
+			left, err = left.ToUint()
+			if err != nil {
+				return uintOp, null, null, err
+			}
+			right, err = right.ToUint()
+			if err != nil {
+				return uintOp, null, null, err
+			}
+			return uintOp, left, right, nil
+		}
+		if left.IntVal != nil && right.IntVal != nil {
+			return intOp, left, right, nil
+		}
+	}
+
+	left, err = left.ToFloat()
+	if err != nil {
+		return floatOp, null, null, err
+	}
+	right, err = right.ToFloat()
+	if err != nil {
+		return floatOp, null, null, err
+	}
+	return floatOp, left, right, nil
+}
+
+func evalBinaryExpr(expr *parser.BinaryExpr, env Env) (sqlwire.Datum, error) {
+	left, err := EvalExpr(expr.Left, env)
+	if err != nil {
+		return null, err
+	}
+	right, err := EvalExpr(expr.Right, env)
+	if err != nil {
+		return null, err
+	}
+
+	// TODO(pmattis): Overflow/underflow checks?
+
+	var typ opType
+
+	switch expr.Operator {
+	case parser.Bitand:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal &= *right.UintVal
+		case intOp:
+			*left.IntVal &= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Bitor:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal |= *right.UintVal
+		case intOp:
+			*left.IntVal |= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Bitxor:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal ^= *right.UintVal
+		case intOp:
+			*left.IntVal ^= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Plus:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
```

Comment:
small comment explaining why which op type is put into `prepareBinaryArgs` would be good.